[JIT] ShapeProp: add missing ops from mobilenet v3.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59347 [TensorExpr] Add NNC lowerings for `aten::mean`, `aten::addmm`, and `aten::adaptive_avg_pool2d`.
* #59346 [TensorExpr] Handle not-specified dtypes and strides.
* **#59163 [JIT] ShapeProp: add missing ops from mobilenet v3.**

Differential Revision: [D28853833](https://our.internmc.facebook.com/intern/diff/D28853833)

Commit:
[JIT] ShapeProp: add missing ops from mobilenet v3.

[ghstack-poisoned]

Commit:
Update on "[JIT] ShapeProp: add missing ops from mobilenet v3."




[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853833).

desertfire commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -49,6 +49,39 @@ const std::string shape_compute_functions =
+          return out
+
+        def view(self: List[int], sizes: List[int]):
+          out: List[int] = []
```
Do we need to add some assertion here?

Commit:
Update on "[JIT] ShapeProp: add missing ops from mobilenet v3."


Differential Revision: [D28853833](https://our.internmc.facebook.com/intern/diff/D28853833)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853833).

eellison commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -49,6 +49,43 @@ const std::string shape_compute_functions =
+            out.append(elem)
+          return out
+
+        def mean_dim(self: List[int], dims: List[int], keep_dim: bool, dt : Any):
```
where is this code taken from ? it doesn't look that similar to the logic i find in eager: https://codebrowser.bddppq.com/pytorch/pytorch/aten/src/ATen/native/ReduceOpsUtils.h.html#_ZN2at6native23create_reduction_resultERKNS_6TensorEN3c108ArrayRefIlEEbNS4_10ScalarTypeE

ZolotukhinM replied:
I wrote it from scratch, do you think it's different from this one?
https://github.com/pytorch/pytorch/blob/1798ff02e430ad4c9ca1aff218913142da39d4a8/aten/src/ATen/native/ReduceOpsUtils.h#L137-L145
I didn't explicitly create the mask like the code for the eager mode, but i'm not sure it's different.

eellison replied:
it might be ... but we can take it as a follow up to replace this with the implementation logic, as with `view`. it's much easier to reason about when there is a direct 1-1 correspondence... 

future PRs we can be more strict. adding OpInfo tests (which i'm working on) should also help flush out partially-covered semantics

eellison replied:
upon further comparison LGTM...

eellison commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -49,6 +49,43 @@ const std::string shape_compute_functions =
+          # TODO: add assertions to check whether requested dims are valid
+          out: List[int] = []
+          for elem in sizes:
+            if elem == -1:
```
i guess i'm fine unblocking this, but adding an assertion doesn't actually prevent the analysis from working. we would also have to add handling in the analysis so that if an exception is on the graph block we don't set the output. in general i would like to stay away from only partially handling the behavior of an operator; in the long-term i don't think it's much simpler than just adding complete support.

eellison reviewed (approved):
fine landing to unblock but we should fix the things mentioned in my comments before adding any other users of shape analysis...

Commit:
Update on "[JIT] ShapeProp: add missing ops from mobilenet v3."


Differential Revision: [D28853833](https://our.internmc.facebook.com/intern/diff/D28853833)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853833).

Commit:
rebase on "[JIT] ShapeProp: add missing ops from mobilenet v3."


Differential Revision: [D28853833](https://our.internmc.facebook.com/intern/diff/D28853833)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853833).

Commit:
rebase on "[JIT] ShapeProp: add missing ops from mobilenet v3."


Differential Revision: [D28853833](https://our.internmc.facebook.com/intern/diff/D28853833)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853833).

Commit:
rebase on "[JIT] ShapeProp: add missing ops from mobilenet v3."


Differential Revision: [D28853833](https://our.internmc.facebook.com/intern/diff/D28853833)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853833).

Commit:
Update on "[JIT] ShapeProp: add missing ops from mobilenet v3."


Differential Revision: [D28853833](https://our.internmc.facebook.com/intern/diff/D28853833)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853833).

Commit:
rebase on "[JIT] ShapeProp: add missing ops from mobilenet v3."


Differential Revision: [D28853833](https://our.internmc.facebook.com/intern/diff/D28853833)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28853833).
